keps/0019-package-api-versioning.md

```diff
@@ -2,64 +2,374 @@
+
+##### Cons
+
+- Operator users care about the application version, not a "synthetic" version
+- Makes it complex to have an "app version"-based backport strategy. Customers might not be able to use operator based on newer app versions due to constraints outside of our control, but would still want to receive new operator-related features and bug fixes
+- Continuing from the second "pro", it might be possible that subsequent package versions with different app versions don't contain an operator-related feature or bug fix that only makes sense for a specific app version (e.g., version N+1 based on app version M doesn't include an operator feature available on version N based on app version M+1 because the operator feature only makes sense for app version M)
+- The progression of the package version is confusing with regards to the application version progression
+
+  Example:
+
+  | Time | Package/operator version | App version        | Comment                                            |
+  | ---- | ---------------          | ------------------ | -------------------------------------------------- |
+  | T0   | 1.0.0                    | 2.3.0              | Initial release based on app version 2.3.x         |
+  | T1   | 1.1.0                    | 3.0.0              | Release based on app version 3.0.x                 |
+  | T2   | 1.2.0                    | 2.3.0              | New operator-related feature for app version 2.3.x |
+  | T2   | 1.3.0                    | 3.0.0              | New operator-related feature for app version 3.0.x |
+  | T3   | 1.4.0                    | 2.3.1              | Bug fix release for app version 2.3.x              |
+  | T4   | 1.5.0                    | 2.3.1              | Important bug fix A in operator-related code       |
+  | T4   | 1.6.0                    | 3.0.0              | Important bug fix A in operator-related code       |
+  | T5   | 1.6.1                    | 2.3.1              | Small bug fix B in operator-related code           |
+  | T5   | 1.6.2                    | 3.0.0              | Small bug fix B in operator-related code           |
+
+  The example above illustrates two sources of confusion:
+  - A higher package version can contain a lower app version
+  - There's no relation between the package version growth and the app version growth
+
+#### Composite package/operator version
+
+##### Pros
+
+- Puts the app version (what the user cares about) front and center
+- Makes it trivial to have an "app version"-based backport/release strategy
+- Makes it trivial to have an "app version"-based support strategy (e.g., an organization communicates official support for only the N-2 versions of an operator)
+
+##### Cons
+
+- Visually more complex
+- Makes it complex to know which operator-related features are available in which versions
+
+##### Example hypothetical timeline for Apache Cassandra releases
+
+Remember that _version_ is _operator version_ is _package version_.
```

Comment:
is there a typo in this sentence?

Reply:
Not an intentional one at least... I thought `A is B is C` was an idiom to express `A = B = C` 😛

Reply:
it gets really confusing here as operator version isn't operator revision. So I thought we were missing a `hyphen` in above sentence